tools/fill_db.py

```diff
@@ -0,0 +1,276 @@
+        v = np.searchsorted(distMap, u)
+
+        samples = [t - 1 for t in v]
+        yield samples
+
+
+def rearrange_keys():
+    """
+    This function potentially provides the capability for testing different caching workloads.
+    for instance, if we rearrange all the keys via sorting based on the k-v memory usage,
+    we will generate a zipfian hotspot that prefers to access small kv pairs (or larger kv pairs)
+    current implementation just uses a random shuffle.
+    """
+    global inserted_keys
+    shuffle(inserted_keys)
+
+
+async def query_db_with_locality(count):
+    global inserted_keys
+    r = aioredis.Redis(decode_responses=True)
+    n = 0
+    read_keys()
+    rearrange_keys()
+    misses = 0
+    hits = 0
+    pipeline_size = 200
+    key_index_gen = rand_zipf_generator(1.0, len(inserted_keys), pipeline_size)
+    for key_indices in key_index_gen:
+        pipeline = r.pipeline(False)
+        # print(key_indices)
+        for key_index in key_indices:
+            k = inserted_keys[key_index]
+            pipeline.set(k, random_val(), nx=True)
+
+        responses = await pipeline.execute()
+        n += pipeline_size
+        for resp in responses:
+            if resp:
+                misses += 1
+            else:
+                hits += 1
+        print(
+            "\r>> Number of ops: {0}, hit: {1}, miss: {2}, hit rate: {3:.4f}".format(
+                n, hits, misses, float(hits) / float(hits + misses)
+            ),
+            end="",
+        )
+        if n >= count:
+            break
+    hit_rate = float(hits) / float(hits + misses)
+    print("\n>> Cache hit rate: {:.4f}".format(hit_rate))
+
+
+class Range(object):
+    def __init__(self, start, end):
+        self.start = start
+        self.end = end
+
+    def __eq__(self, other):
+        return self.start <= other <= self.end
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser(
+        description="Cache Benchmark", formatter_class=argparse.ArgumentDefaultsHelpFormatter
+    )
+
+    parser.add_argument(
+        "-f",
+        "--fill",
+        action="store_true",
+        help="fill database with random key-value pairs with their lengths follow some distributions",
+    )
+
+    parser.add_argument(
+        "-r",
+        "--ratio",
+        type=float,
+        default=0.5,
+        choices=[Range(0.0, 1.0)],
+        help="the ratio between in memory data size and total data size",
+    )
+
+    parser.add_argument(
+        "-m",
+        "--measure",
+        action="store_true",
+        help="measure cache hit rate by visiting the entire key space with a Zipfian distribution",
+    )
+
+    parser.add_argument(
+        "-c",
+        "--count",
+        type=int,
+        default=100000,
+        help="total number of operations to be performed when measuring cache hit rate",
+    )
+
+    args = parser.parse_args()
+
+    if args.fill:
+        asyncio.run(populate_db(args.ratio))
+        exit(0)
```

Comment:
> If Python is not 100% bottlenecked on CPU due to zipfan calcuations, instead of starting multiple processes you can run multiple asyncio tasks that would parallelize io
> 

Yes, I agree. I think this is also doable. I simply did not try it, but I think it should work.

> I also remember we had a golang tool that tested cache efficiency with zipfan

I wasn't aware of the golang tool, maybe @ashotland knows? I'm aware of the cache test python tool @romange created in the tool directory, my zipf generator is stolen from his. :-)

Reply:
[It's here](https://github.com/dragonflydb/playground/tree/main/apps/zipfantest)